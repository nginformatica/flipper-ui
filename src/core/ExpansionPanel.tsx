import MuiExpansionPanel from '@material-ui/core/ExpansionPanel'
import MuiExpansionPanelActions from '@material-ui/core/ExpansionPanelActions'
import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import React, { ReactNode, FC, MouseEvent } from 'react'
import { IProps as IPaper } from './Paper'
import {
    HelperBox,
    TextFieldWrapper as ExpansionPanelHeaderWrapper
} from './TextField'

interface IProps extends IPaper {
    actions?: ReactNode
    defaultExpanded?: boolean
    details?: ReactNode
    disabled?: boolean
    expandIcon?: ReactNode
    expanded?: boolean
    summary?: ReactNode
    summaryStyle?: object
    detailsStyle?: object
    actionsStyle?: object
    helperIcon?: React.ReactNode
    helperButtonPosition?: 'left' | 'right'
    onHelperClick?: () => void
    onChange?: (event?, expanded?) => void
}

const ExpansionPanel: FC<IProps> = ({
    actions,
    details,
    expandIcon,
    margin,
    padding,
    style,
    summary,
    summaryStyle,
    detailsStyle,
    actionsStyle,
    onHelperClick,
    helperIcon,
    helperButtonPosition = 'right',
    ...otherProps
}) => {

    const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
        if (onHelperClick) {
            event.stopPropagation()
            onHelperClick()
        }
    }

    const renderHelper = (
        <>
            {
                onHelperClick && (
                    <HelperBox
                        helperIcon
                        onHelperClick={ handleClick }
                    />
                )
            }
        </>
    )

    return (
        <MuiExpansionPanel
            { ...otherProps }
            style={ { margin, padding, ...style } }>
            {
                summary && (
                    <MuiExpansionPanelSummary
                        expandIcon={ expandIcon }
                        style={ summaryStyle }>
                        <ExpansionPanelHeaderWrapper>
                            { helperButtonPosition === 'left' && renderHelper }
                            { summary }
                            { helperButtonPosition === 'right' && renderHelper }
                        </ExpansionPanelHeaderWrapper>
                    </MuiExpansionPanelSummary>
                )
            }
            {
                details && (
                    <MuiExpansionPanelDetails style={ detailsStyle }>
                        { details }
                    </MuiExpansionPanelDetails>
                )
            }
            {
                actions && (
                    <MuiExpansionPanelActions style={ actionsStyle }>
                        { actions }
                    </MuiExpansionPanelActions>
                )
            }
        </MuiExpansionPanel >
    )
}

export default ExpansionPanel
