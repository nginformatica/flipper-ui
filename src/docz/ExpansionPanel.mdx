---
name: Expansion Panel
route: /expansion-panel
---

import Button from '../core/Button'
import ExpansionPanel from '../core/ExpansionPanel'
import Typography from '../core/Typography'
import ExpandMore from '@material-ui/icons/ExpandMore'
import { Playground, Props } from 'docz'

# ExpansionPanel

## Properties
<Props of={ ExpansionPanel } />

## Demo
<Playground>
    <ExpansionPanel
        summary='Pulp Fiction'
        details={
            <div>
                <Typography gutterBottom>
                    books about imaginary characters and events, produced in large quantities
                    and intended to be read by many people but not considered to be of very good quality:
                </Typography>
                <Typography variant='caption'>
                    She was reading a pulp fiction thriller with a lurid cover.
                </Typography>
            </div>
        }
        onHelperClick={ () => window.alert('HELP!') }
        expandIcon={ <ExpandMore /> }
        actions={
            <Button color='primary'>
                Confirm
            </Button>
        }
    />
    <ExpansionPanel
        summary='Mentalist'
        details='Someone who uses mental acuity, hypnosis and/or suggestion.'
        expandIcon={ <ExpandMore /> }
    />
    <ExpansionPanel
        disabled
        summary='Dormammu'
        details='I have come to bargain'
        expandIcon={ <ExpandMore /> }
    />
</Playground>
