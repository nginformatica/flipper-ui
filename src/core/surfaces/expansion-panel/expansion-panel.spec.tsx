import * as React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import ExpansionPanel from '.'

const expectToThrow = (fn: () => void, message: string) => {
    // Even though the error is caught, it still gets printed to the console
    // so we mock that out to avoid the wall of red text.
    const spy = jest.spyOn(console, 'error')
    spy.mockImplementation(jest.fn())
    expect(fn).toThrow(message)
    spy.mockRestore()
}

describe('ExpansionPanel', () => {
    it('should render closed', () => {
        render(
            <ExpansionPanel
                expanded={false}
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onClick={jest.fn()}
            />
        )

        const expansionPanel = screen.getByRole('mui-expansion-panel')
        expect(expansionPanel.lastChild).toHaveProperty(
            'className',
            'MuiCollapse-root MuiCollapse-hidden'
        )
    })

    it('should render expanded', () => {
        render(
            <ExpansionPanel
                expanded
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onClick={jest.fn()}
            />
        )

        const expansionPanel = screen.getByRole('mui-expansion-panel')
        expect(expansionPanel.lastChild).toHaveProperty(
            'className',
            'MuiCollapse-root MuiCollapse-entered'
        )
    })

    it('should render with actions', () => {
        render(
            <ExpansionPanel
                expanded
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onClick={jest.fn()}
                actions={<button role='action-button'>Confirm</button>}
            />
        )

        const expansionPanel = screen.getByRole('action-button')
        expect(expansionPanel.parentElement?.classList).toContain(
            'MuiAccordionActions-root'
        )
    })

    it('should render helper box on left', () => {
        const onHelperClick = jest.fn()
        render(
            <ExpansionPanel
                expanded
                helperButtonPosition='left'
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onHelperClick={onHelperClick}
                onClick={jest.fn()}
            />
        )

        const helperIcon = screen.getByRole('helper-box')

        expect(helperIcon.nextSibling).toBeTruthy()
    })

    it('should call onHelperClick', async () => {
        const onHelperClick = jest.fn()
        render(
            <ExpansionPanel
                expanded
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onHelperClick={onHelperClick}
                onClick={jest.fn()}
            />
        )

        const helperIcon = await screen.findByRole('helper-box')

        fireEvent.click(helperIcon.firstChild || helperIcon)

        await waitFor(() => expect(onHelperClick).toHaveBeenCalled())
    })

    it('should call onEditClick', async () => {
        const onEditClick = jest.fn()
        render(
            <ExpansionPanel
                expanded
                editable
                editing={false}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onEditClick={onEditClick}
                onClick={jest.fn()}
                onSaveClick={jest.fn()}
                onHelperClick={jest.fn()}
            />
        )

        const editBoxContainer = await screen.findByRole('edit-box')

        fireEvent.click(editBoxContainer.firstChild || editBoxContainer)

        await waitFor(() => expect(onEditClick).toHaveBeenCalled())
    })

    it('should call onSaveClick', async () => {
        const onSaveClick = jest.fn()
        render(
            <ExpansionPanel
                expanded
                editable
                editing={true}
                summary='Expansion Panel'
                role='mui-expansion-panel'
                onEditClick={jest.fn()}
                onClick={jest.fn()}
                onSaveClick={onSaveClick}
                onHelperClick={jest.fn()}
            />
        )

        const editBoxContainer = await screen.findByRole('edit-box')

        fireEvent.click(editBoxContainer.firstChild || editBoxContainer)

        await waitFor(() => expect(onSaveClick).toHaveBeenCalled())
    })

    it('should throw if onSaveClick is not defined on edit mode', () => {
        expectToThrow(() => {
            render(
                <ExpansionPanel
                    expanded
                    editing
                    summary='Expansion Panel'
                    role='mui-expansion-panel'
                    onClick={jest.fn()}
                    actions={<button role='action-button'>Confirm</button>}
                    editable
                />
            )
        }, 'onSaveClick is required when editable is true')
    })

    it('should throw if onEditClick is not defined on edit mode', () => {
        expectToThrow(() => {
            render(
                <ExpansionPanel
                    expanded
                    editing
                    summary='Expansion Panel'
                    role='mui-expansion-panel'
                    onSaveClick={jest.fn()}
                    onClick={jest.fn()}
                    actions={<button role='action-button'>Confirm</button>}
                    editable
                />
            )
        }, 'onEditClick is required when editable is true')
    })
})
