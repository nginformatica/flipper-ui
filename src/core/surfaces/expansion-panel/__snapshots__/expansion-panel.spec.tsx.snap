// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpansionPanel should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters css-1086bdv-MuiPaper-root-MuiAccordion-root"
    role="mui-expansion-panel"
  >
    <div
      aria-expanded="true"
      class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-1f773le-MuiButtonBase-root-MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-eqpfi5-MuiAccordionSummary-content"
      >
        <div
          class="sc-iGgVNO iPPrNl"
        >
          Expansion Panel
        </div>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            class="MuiAccordion-region"
            role="region"
          >
            <div
              class="MuiAccordionActions-root MuiAccordionActions-spacing css-adcfpp-MuiAccordionActions-root"
            >
              <button
                role="action-button"
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
